lib/vrl/stdlib/src/mod_func.rs

```diff
@@ -0,0 +1,106 @@
+use ::value::Value;
+use vrl::prelude::*;
+
+fn r#mod(value: Value, modulus: Value) -> Resolved {
+    let result = value.try_rem(modulus)?;
+    Ok(result)
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Mod;
+
+impl Function for Mod {
+    fn identifier(&self) -> &'static str {
+        "mod"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::INTEGER | kind::FLOAT,
+                required: true,
+            },
+            Parameter {
+                keyword: "modulus",
+                kind: kind::INTEGER | kind::FLOAT,
+                required: true,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "mod",
+            source: r#"mod(5, 3)"#,
+            result: Ok("2"),
+        }]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::TypeState,
+        _ctx: &mut FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let modulus = arguments.required("modulus");
```

Comment:
One addition we can add here, is checking if `modulus` is a literal 0, and then error at compile-time.

The chances of that happening are minimal, so I also wouldn't mind leaving this for a possible future enhancement opportunity (perhaps, if you're applying some code-review changes anyway, you could at least note this as a future improvement in an inline comment)

Reply:
I'll add a comment, but my intention was to keep this implementation as close as possible to the previous `modulo` operator to make it easier to migrate. We have a lot of these types of issues that can be caught at runtime and I'd like to handle them in a slightly more general way, and then have some kind of audit to go through functions and improve them all together.
I made an issue for it already here: https://github.com/vectordotdev/vector/issues/13968